src/lib.rs

```diff
@@ -1198,36 +1174,68 @@ decl_module! {
-        //     match thread.poll {
-        //         None => Err(ERROR_POLL_NOT_EXIST),
-        //         Some(poll) => {
-        //             if poll.end_time < <timestamp::Module<T>>::now() {
-        //                 Err(ERROR_POLL_COMMIT_EXPIRED)
-        //             } else {
-        //                 Self::ensure_poll_data_valid(&poll, poll_value)?;
-        //                 <PollDataByAccount<T>>::insert(thread_id, forum_user_id, poll_value);
-        //                 if <PollStatistics<T>>::exists(thread_id, poll_value) {
-        //                     <PollStatistics<T>>::insert(thread_id, poll_value, 1);
-        //                 } else {
-        //                     let old_value = <PollStatistics<T>>::get(thread_id, poll_value);
-        //                     <PollStatistics<T>>::insert(thread_id, poll_value, old_value + 1);
-        //                 }
-        //                 Ok(())
-        //             }
-        //         }
-        //     }
-        // }
+        fn submit_poll(origin, forum_user_id: T::ForumUserId, thread_id: T::ThreadId, index: u32) -> dispatch::Result {
+            // Check that its a valid signature
+            let who = ensure_signed(origin)?;
+
+            // get forum user id.
+            Self::ensure_is_forum_member_with_correct_account(&who, &forum_user_id)?;
+
+            // Get thread
+            let thread = Self::ensure_thread_exists(&thread_id)?;
+
+            // Make sure poll exist
+            match thread.poll {
+                None => Err(ERROR_POLL_NOT_EXIST),
+                Some(poll) => {
+                    // Poll not expired
+                    if poll.end_time < <timestamp::Module<T>>::now() {
+                        Err(ERROR_POLL_COMMIT_EXPIRED)
+                    } else {
+                        let alternative_length = poll.poll_alternatives.len();
+                        // The selected alternative index is valid
+                        if index as usize >= alternative_length {
+                            Err(ERROR_POLL_DATA)
+                        } else {
+                            // Update thread data
+                            <ThreadById<T>>::mutate(thread_id, |value| {
+                                // Store new poll alternative statistics
+                                let mut new_vote_count = poll.poll_alternatives.clone();
+
+                                new_vote_count[index as usize] = PollAlternative{
+                                    alternative_text: new_vote_count[index as usize].alternative_text.clone(),
+                                    // Increment one
+                                    poll_count: new_vote_count[index as usize].poll_count + 1,
+                                };
+
+                                // Update thread with one object
+                                *value = Thread {
+                                    id: value.id,
```

Comment:
Use struct update syntax, rather than specify from scratch.

https://doc.rust-lang.org/stable/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax

Reply:
the grammar is helpful, accepted.